state/prunning_service.go

```diff
@@ -0,0 +1,228 @@
+package state
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/cometbft/cometbft/libs/service"
+)
+
+type PruningRequester int64
+
+const (
+	AppRequester                 PruningRequester = 0
+	DataCompanionRequester       PruningRequester = 1
+	ABCIResRetainHeightRequester PruningRequester = 2
+)
+
+var (
+	AppRetainHeightKey           = []byte("AppRetainHeightKey")
+	DataCompanionRetainHeightKey = []byte("DCRetainHeightKey")
+	ABCIResultsRetainHeightKey   = []byte("ABCIResRetainHeightKey")
+)
+
+type Pruner struct {
+	service.BaseService
+	logger log.Logger
+	// DB to which we save the retain heights
+	bs BlockStore
+	// State store to prune state from
+	stateStore    Store
+	pruningTicker *time.Ticker
+}
+
+type RetainHeightInfo struct {
+	Height    int64
+	Requester PruningRequester
+}
+
+func NewPruner(stateStore Store, bs BlockStore, logger log.Logger) *Pruner {
+	p := &Pruner{
+		bs:            bs,
+		stateStore:    stateStore,
+		logger:        logger,
+		pruningTicker: time.NewTicker(time.Second * 10),
+	}
+	p.BaseService = *service.NewBaseService(logger, "Prunning service", p)
+	return p
+}
+
+// If OnStart() returns an error, it's returned by Start()
+func (p *Pruner) OnStart() error {
+	if err := p.BaseService.OnStart(); err != nil {
+		return err
+	}
+	go p.pruningRoutine()
+	return nil
+}
+
+// Stop the service.
+// If it's already stopped, will return an error.
+// OnStop must never error.
+func (p *Pruner) OnStop() {
+	p.pruningTicker.Stop()
+	p.BaseService.OnStop()
+
+}
+
+// SetPruningHeight is called by either the application or the data companion to set the
+// retain height for blocks or ABCI block results.
+// It returns the height which will be used as the retain height and an error message in case
+// the retain height cannot be set to the desired value (if the data was already pruned, if a lower
+// retain height was already set)
+func (p *Pruner) SetPruningHeight(retainHeightInfo RetainHeightInfo) (retainHeight int64, err error) {
```

Comment:
I'd recommend splitting this method into 3 different ones to reflect the business logic better at compile-time (the current approach effectively does this at runtime, making it more difficult to test):
1. `SetApplicationRetainHeight` to handle the `AppRequester` part of the `switch` statement
2. `SetCompanionBlockRetainHeight` (to be specific about the fact that it influences block retention) to handle the `DataCompanionRequester` part of the `switch` statement
3. `SetABCIResRetainHeight` to handle the `ABCIResRetainHeightRequester` part of the `switch` statement

A general rule of thumb is that each function should only do one thing (and ideally only 20-30 lines of code per function so it's quick and easy to read - not always possible, of course, but something to aim for).